worker/agentconfigupdater/worker.go

```diff
@@ -0,0 +1,120 @@
+// Copyright 2019 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package agentconfigupdater
+
+import (
+	"time"
+
+	"github.com/juju/errors"
+	"github.com/juju/pubsub"
+	"gopkg.in/juju/worker.v1"
+	"gopkg.in/tomb.v2"
+
+	coreagent "github.com/juju/juju/agent"
+	"github.com/juju/juju/mongo"
+	controllermsg "github.com/juju/juju/pubsub/controller"
+	jworker "github.com/juju/juju/worker"
+)
+
+// WorkerConfig contains the information necessary to run
+// the agent config updater worker.
+type WorkerConfig struct {
+	Agent        coreagent.Agent
+	Hub          *pubsub.StructuredHub
+	MongoProfile mongo.MemoryProfile
+	Logger       Logger
+}
+
+// Validate ensures that the required values are set in the structure.
+func (c *WorkerConfig) Validate() error {
+	if c.Agent == nil {
+		return errors.NotValidf("missing agent")
+	}
+	if c.Hub == nil {
+		return errors.NotValidf("missing hub")
+	}
+	if c.Logger == nil {
+		return errors.NotValidf("missing logger")
+	}
+	return nil
+}
+
+type agentConfigUpdater struct {
+	config WorkerConfig
+
+	tomb         tomb.Tomb
+	mongoProfile mongo.MemoryProfile
+}
+
+// NewWorker creates a new agent config updater worker.
+func NewWorker(config WorkerConfig) (worker.Worker, error) {
+	if err := config.Validate(); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	started := make(chan struct{})
+	w := &agentConfigUpdater{
+		config:       config,
+		mongoProfile: config.MongoProfile,
+	}
+	w.tomb.Go(func() error {
+		return w.loop(started)
+	})
+	select {
+	case <-started:
+	case <-time.After(10 * time.Second):
+		return nil, errors.New("worker failed to start properly")
+	}
```

Comment:
this isn't something we typically do in NewWorker functions. Is it something you want to see start getting added?

Reply:
Particularly for workers that also do initiialization in the main loop, like those that subscribe to events, it is required to ensure that the worker has fully set up before we return otherwise there are racy tests.